flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java

```diff
@@ -0,0 +1,259 @@
+ *         and 2) sens a custom {@link SourceEvent SourceEvents} to the source readers.
+ *     </li>
+ *     <li>
+ *         Thread model enforcement - The context ensures that all the manipulations to the coordinator state
+ *         are handled by the same thread.
+ *     </li>
+ * </ul>
+ * @param <SplitT> the type of the splits.
+ */
+@Internal
+public class SourceCoordinatorContext<SplitT extends SourceSplit> implements SplitEnumeratorContext<SplitT> {
+	private final ExecutorService coordinatorExecutor;
+	private final ExecutorNotifier notifier;
+	private final OperatorCoordinator.Context operatorCoordinatorContext;
+	private final ConcurrentMap<Integer, ReaderInfo> registeredReaders;
+	private final SplitAssignmentTracker<SplitT> assignmentTracker;
+	private final String coordinatorThreadName;
+
+	public SourceCoordinatorContext(
+			ExecutorService coordinatorExecutor,
+			String coordinatorThreadName,
+			int numWorkerThreads,
+			OperatorCoordinator.Context operatorCoordinatorContext) {
+		this(coordinatorExecutor, coordinatorThreadName, numWorkerThreads, operatorCoordinatorContext,
+				new SplitAssignmentTracker<>());
+	}
+
+	// Package private method for unit test.
+	SourceCoordinatorContext(
+			ExecutorService coordinatorExecutor,
+			String coordinatorThreadName,
+			int numWorkerThreads,
+			OperatorCoordinator.Context operatorCoordinatorContext,
+			SplitAssignmentTracker<SplitT> splitAssignmentTracker) {
+		this.coordinatorExecutor = coordinatorExecutor;
+		this.notifier = new ExecutorNotifier(
+				Executors.newScheduledThreadPool(numWorkerThreads, new ThreadFactory() {
+					private int index = 0;
+					@Override
+					public Thread newThread(Runnable r) {
+						return new Thread(r, coordinatorThreadName + "-worker-" + index++);
+					}
+				}),
+				coordinatorExecutor);
+		this.operatorCoordinatorContext = operatorCoordinatorContext;
+		this.registeredReaders = new ConcurrentHashMap<>();
+		this.assignmentTracker = splitAssignmentTracker;
+		this.coordinatorThreadName = coordinatorThreadName;
+	}
+
+	@Override
+	public MetricGroup metricGroup() {
+		return null;
+	}
+
+	@Override
+	public void sendEventToSourceReader(int subtaskId, SourceEvent event) {
+		try {
+			operatorCoordinatorContext.sendEvent(new SourceEventWrapper(event), subtaskId);
+		} catch (TaskNotRunningException e) {
+			throw new FlinkRuntimeException(String.format("Failed to send event %s to subtask %d",
+					event,
+					subtaskId), e);
+		}
+	}
+
+	@Override
+	public int currentParallelism() {
+		return operatorCoordinatorContext.currentParallelism();
+	}
+
+	@Override
+	public Map<Integer, ReaderInfo> registeredReaders() {
+		return Collections.unmodifiableMap(registeredReaders);
+	}
+
+	@Override
+	public void assignSplits(SplitsAssignment<SplitT> assignment) {
+		// Ensure the split assignment is done by the the coordinator executor.
+		if (!Thread.currentThread().getName().equals(coordinatorThreadName)) {
```

Comment:
Going by thread name is very fragile. Why not go by reference equality of the Thread objects?